File: 2023/Rust/aoc_2023/src/day_1.rs
use std::fs;
use regex::Regex; 

pub fn day_1_main() {
    //Ingest data
    let file_name = "inputs/day_1_input.txt";
    let contents = fs::read_to_string(file_name)
        .expect("Failed to read File");

    //Split by each line
    let parts: Vec<&str> = contents.split("\n").collect();

    println!("The answer the Part 1 is: {}", part_1(parts.clone()));
    println!("The answer the Part 2 is: {}", part_2(parts.clone()));
}

fn part_1(parts: Vec<&str>) -> i32 {
    let mut sumr = 0;

    for i in 0..parts.len()-1{
        let mut line: Vec<&str> = parts[i].split("").collect();

        //First Number
        let first_num = get_number(line.clone());

        line.reverse();

        //Second Number
        let second_num = get_number(line.clone());

        let final_num:i32 = (first_num.to_owned() + "" + second_num).parse().unwrap();

        sumr = sumr + final_num;
    }

    return sumr;
}

fn part_2(parts: Vec<&str>) -> i32 {
    let num_spelled = ["one","two","three","four","five","six","seven","eight","nine"];
    let nums = ["1","2","3","4","5","6","7","8","9"];
    let mut sumr = 0;

    for i in 0..parts.len()-1{
        let mut parts_line: String = parts[i].to_string(); 

        
        for j in 0..num_spelled.len(){
            //Need to add first letter of word and last letter of word around the number
            let mut replace_val: String = num_spelled[j].chars().next().unwrap().to_string();
            replace_val = replace_val + nums[j]; 
            replace_val = replace_val + &num_spelled[j].chars().last().unwrap().to_string();

            parts_line = parts_line.replace(num_spelled[j], &replace_val);
        }
        let mut line: Vec<&str> = parts_line.split("").collect();
        //println!("{:?}", line);

        //First Number
        let first_num = get_number(line.clone());

        line.reverse();

        //Second Number
        let second_num = get_number(line.clone());
        let final_num:i32 = (first_num.to_owned() + "" + second_num).parse().unwrap();

        sumr = sumr + final_num;
    }

    return sumr;
}

fn get_number(line: Vec<&str>) -> &str{
    let num_regex = Regex::new(r"[0-9]").unwrap();
    for i in 1..line.len()-1{
        let item_check = line[i];

        if num_regex.is_match(item_check){
            return item_check;
        }
    }

    return "";
}

File: 2023/Python/day_1.py
import re


def parse_input():
    splitter = []
    with open("Inputs/day_1_input.txt") as input_file:
        for line in input_file:
            ltext = line.replace("\n", "")
            splitter.append(ltext)
    return splitter


def part_1():
    data = parse_input()
    sumr = 0

    for i in data:
        num_arr = []
        l_list = list(i)

        for j in l_list:
            if re.search(r"[0-9]", j):
                num_arr.append(j)
        first_num = num_arr[0]
        last_num = num_arr[len(num_arr) - 1]
        sumr += int(first_num + last_num)

    print("The sum for Part 1 is: " + str(sumr))


def part_2():
    data = parse_input()
    sumr = 0

    nums_spelled = [
        "one",
        "two",
        "three",
        "four",
        "five",
        "six",
        "seven",
        "eight",
        "nine",
    ]
    nums = ["1", "2", "3", "4", "5", "6", "7", "8", "9"]

    for i in data:
        print("Pre : " + i)
        for x in range(len(nums_spelled)):
            i = i.replace(
                nums_spelled[x],
                (
                    nums_spelled[x][0]
                    + nums[x]
                    + nums_spelled[x][
                        len(nums_spelled[x]) - 1
                    ]  # Need to add first and letter to the replace, fiveeight should be 58, not 5ight
                ),
            )
        print("Post: " + i)

        num_arr = []
        l_list = list(i)

        for j in l_list:
            if re.search(r"[0-9]", j):
                num_arr.append(j)
        first_num = num_arr[0]
        last_num = num_arr[len(num_arr) - 1]
        sumr += int(first_num + last_num)

    print("The sum for Part 2 is: " + str(sumr))


if __name__ == "__main__":
    part_1()
    part_2()
